### DISCUSSION TITLE
How to implement Sepia filter?

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#27] [Q&A] [answered]
Discuss algorithm on how to apply `sepia` effect on `BMP` image.

**Comment by chandrakant100** [comment]
| Original Image and its Sepia image |
| ---------------------------------------- |
| <img src="https://cms-assets.tutsplus.com/cdn-cgi/image/width=850/uploads/users/108/posts/34119/image/sepia-effect-photoshop-final.jpg" height="300px" width="500px"> |

**Reply by Tushar7-coder** [reply]
`sepia` filter gives images an old-timey feel by making the whole image look a bit reddish-brown.

An image can be converted to sepia by taking each pixel, and computing new red, green, and blue values based on the original values of the three.

There are a number of algorithms for converting an image to sepia.
 For each pixel, the sepia color values should be calculated based on the original color values per the below.

  ```
  sepiaRed = .393 * originalRed + .769 * originalGreen + .189 * originalBlue
  sepiaGreen = .349 * originalRed + .686 * originalGreen + .168 * originalBlue
  sepiaBlue = .272 * originalRed + .534 * originalGreen + .131 * originalBlue
```
The result of each of these formulas may not be an integer(floating point value), but each value could be rounded to the nearest integer. It’s also possible that the result of the formula is a number greater than 255, the maximum value for an 8-bit color value. In that case, the red, green, and blue values should be capped at 255. As a result, we can guarantee that the resulting red, green, and blue values will be whole numbers between 0 and 255, inclusive.

**Reply by H1manshu21** [reply]
Transfer this reply to the `suggest an answer` section so that the answer can be accepted.

**Reply by H1manshu21** [reply]
Can anyone write a pseducode on how to implement to sepia?

**Reply by SHembram18** [reply]
### In Sepia filter , When we apply this filter to an image it converts to Reddish - Brown type image .

_@Tushar7-coder gives the algorthim for applying the filter._

> I have a doubt if the result of the RGB channel excced the limit 255 , At that time how can we handle it?

please anyone answer me

**Reply by Ank1taS** [reply]
> ### In the Sepia filter, When we apply this filter to an image it converts to a Reddish-Brown type image.
> _@Tushar7-coder gives the algorithm for applying the filter._
> 
> > I have a doubt if the result of the RGB channel exceeds the limit 255, At that time how can we handle it?
> 
> please anyone answer me

@HembramBeta777 It is already mentioned that if either of the channel value exceeds 255, then that channel will be assigned 255 i.e max value

**Reply by chandrakant100** [reply]
**`NOTE`**
we cannot use floating point values (if used results will be different).
So the values must be rounded to its closing integer,.
As @Ank1taS said for any value exceeding `255` (max value) we need to assign it to `255`.
(Already mentioned above by @Tushar7-coder )

**Reply by H1manshu21** [reply]
@Tushar7-coder Can you add your previously mentioned Sepia answer in a new section rather than the reply section so that I can mark this answer as accepted, please?

**Comment by chandrakant100** [comment]
## Pseudocode for Sepia Filter
```
function sepia (Argumen1t rows, Argument2 columns, Argument3 pixel 2Darray ) {
        iterate and pick each row of the pixel 2D array
                iterate and pick each pixel of the row picked above
                use 3 different channels to store 3 different colors modification using above formula
                if (valueOfRedchannel or  valueOfGreenchannel or valueOfBluechannel == an floating value) then 
                     round the value to its nearest integer and store it in corresponding channels.
                if (valueOfRedchannel or  valueOfGreenchannel or valueOfBluechannel > 255) then assign it 
                     to 255 and store it in the corresponding channels of 2D array (e.g for valueOfRedchannel 
                     store it in Redchannel of 2D array).
end
}
```

**Comment by Tushar7-coder** [comment, marked as answer]
sepia filter gives images an old-timey feel by making the whole image look a bit reddish-brown.

An image can be converted to sepia by taking each pixel, and computing new red, green, and blue values based on the original values of the three.

There are a number of algorithms for converting an image to sepia.
For each pixel, the sepia color values should be calculated based on the original color values per the below.
```
sepiaRed = .393 * originalRed + .769 * originalGreen + .189 * originalBlue
sepiaGreen = .349 * originalRed + .686 * originalGreen + .168 * originalBlue
sepiaBlue = .272 * originalRed + .534 * originalGreen + .131 * originalBlue
```
(Given by researchers)
The result of each of these formulas may not be an integer(floating point value), but each value could be rounded to the nearest integer. It’s also possible that the result of the formula is a number greater than 255, the maximum value for an 8-bit color value. In that case, the red, green, and blue values should be capped at 255. As a result, we can guarantee that the resulting red, green, and blue values will be whole numbers between 0 and 255, inclusive.
